src/utils/cuda/cuda_mem_component.c

```diff
@@ -27,16 +29,59 @@ static xccl_status_t xccl_cuda_mem_free(void *ptr)
     return XCCL_OK;
 }
 
+static xccl_status_t xccl_cuda_alloc_resources()
+{
+    int i;
+
+    CUDACHECK(cudaStreamCreateWithFlags(&xccl_cuda_mem_component.stream,
+                                        cudaStreamNonBlocking));
+    CUDACHECK(cudaHostAlloc((void**)&xccl_cuda_mem_component.stream_requests,
+              NUM_STREAM_REQUESTS * sizeof(xccl_cuda_mem_component_stream_request_t),
+              cudaHostAllocMapped));
+    for (i = 0; i < NUM_STREAM_REQUESTS; i++) {
+        xccl_cuda_mem_component.stream_requests[i].is_free = 1;
+        CUDACHECK(cudaHostGetDevicePointer(
+                  (void**)&(xccl_cuda_mem_component.stream_requests[i].dev_stop_request),
+                  (void*)&(xccl_cuda_mem_component.stream_requests[i].stop_request),
+                  0));
+        CUDACHECK(cudaEventCreateWithFlags(
+                  &xccl_cuda_mem_component.stream_requests[i].event,
+                  cudaEventDisableTiming));
+    }
+
+    return XCCL_OK;
+}
+
+static xccl_status_t
+xccl_cuda_get_free_stream_request(xccl_cuda_mem_component_stream_request_t **request) {
+    int i;
+    xccl_cuda_mem_component_stream_request_t *req;
+    for (i = 0; i < NUM_STREAM_REQUESTS; i++) {
+        req = xccl_cuda_mem_component.stream_requests + i;
+        if (req->is_free) {
+            req->is_free = 0;
+            *request = req;
+            return XCCL_OK;
+        }
+    }
+
+    return XCCL_ERR_NO_RESOURCE;
+}
+
 xccl_status_t xccl_cuda_reduce_impl(void *sbuf1, void *sbuf2, void *target,
                                     size_t count, xccl_dt_t dtype, xccl_op_t op,
                                     cudaStream_t stream);
 
 xccl_status_t xccl_cuda_reduce(void *sbuf1, void *sbuf2, void *target,
                                size_t count, xccl_dt_t dtype, xccl_op_t op)
 {
+    xccl_status_t st;
+
     if (xccl_cuda_mem_component.stream == 0) {
```

Comment:
Maybe use 1-line macro for the whole this if () {.. } block - it is used multiple times